lib/model/internal_link.dart

```diff
@@ -0,0 +1,285 @@
+import 'package:flutter/foundation.dart';
+import 'package:json_annotation/json_annotation.dart';
+
+import '../api/model/narrow.dart';
+import 'narrow.dart';
+import 'store.dart';
+
+part 'internal_link.g.dart';
+
+const _hashReplacements = {
+  "%": ".",
+  "(": ".28",
+  ")": ".29",
+  ".": ".2E",
+};
+
+final _encodeHashComponentRegex = RegExp(r'[%().]');
+
+// Corresponds to encodeHashComponent in Zulip web;
+// see web/shared/src/internal_url.ts.
+String _encodeHashComponent(String str) {
+  return Uri.encodeComponent(str)
+    .replaceAllMapped(_encodeHashComponentRegex, (Match m) => _hashReplacements[m[0]!]!);
+}
+
+/// Decode a dot-encoded string.
+// The Zulip webapp uses this encoding in narrow-links:
+// https://github.com/zulip/zulip/blob/1577662a6/static/js/hash_util.js#L18-L25
+@visibleForTesting
+String? decodeHashComponent(String str) {
+  try {
+    return Uri.decodeComponent(str.replaceAll('.', '%'));
+  } on ArgumentError {
+    // as with '%1': Invalid argument(s): Truncated URI
+    return null;
+  } on FormatException {
+    // as with '%FF': FormatException: Invalid UTF-8 byte (at offset 0)
```

Comment:
Cool, these comments are helpful — basically documenting in the code the things you learned at https://github.com/zulip/zulip-flutter/pull/305#discussion_r1340487666 , which aren't in upstream's documentation.